contracts/WROSE.sol

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.19;
+
+/*
+ * This WROSE allows meta-withdrawals to accounts that do not have gas.
+ * Token transfers are not private because Sapphire compute node operators can see which accounts' storage slots are active.
+ * Do NOT use this code for anything that requires privacy if you do not trust Sapphire compute node operators.
+ */
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+contract WROSE {
+    using ECDSA for bytes32;
+
+    mapping(address => uint256) private balances;
+    mapping(uint256 => bool) public replayNonce;
+
+    function balanceOf() public view returns (uint256) {
+        return balances[msg.sender];
+    }
+
+    function deposit() public payable {
+        balances[msg.sender] += msg.value;
+    }
+
+    function withdraw(uint256 value) public {
+        require(balances[msg.sender] >= value);
+        balances[msg.sender] -= value;
+        payable(msg.sender).transfer(value);
+    }
+
+    function metaWithdraw(
+        bytes memory signature,
+        address to,
+        uint256 value,
+        uint256 nonce,
+        uint256 reward
+    ) public returns (bool) {
+        bytes32 metaHash = metaWithdrawHash(to, value, nonce, reward);
+        address signer = metaHash.toEthSignedMessageHash().recover(signature);
```

Comment:
I have a strong preference for using typed structure signing as specified in the EIP-712 standard which uses with `eth_signTypedData`.

My concern is that if wrose.io was hijacked by a malicious entity they could change the app to change the `value`, `reward` and `to` parameters behind the scenes, and the user will not be able to identify this change because all they see in the UI is random looking hex encoded string.

![](https://eips.ethereum.org/assets/eip-712/eth_signTypedData.png)

Reply:
You're right. Let me made adjustments for EIP-712.

Very valid points.

Reply:
@HarryR 

Updated to use EIP-712

![Screenshot 2023-05-17 at 4 41 55 PM](https://github.com/wrose-sapphire/contract/assets/19412160/6251d447-bc0b-413e-b531-1c23d82a37d1)